### PR TITLE
Add Bert-style Encoder RFC.

### PR DESCRIPTION
**Pull request by tanzhenyu**
Accepting comment from 09/22/2020 to 10/16/2020.

### CODE REVIEW

File: rfcs/20200920-keras-nlp-bert.md [hunk 1]
```diff
@@ -0,0 +1,369 @@
...
+
+We aim to provide SavedModel for BertEncoder, however user can choose to pretrain a new BertEncoder based on their
+own dataset.
+
+```python
+with strategy.scope():
+  encoder = keras_nlp.encoders.BertEncoder(vocab_size, max_sequence_length, type_vocab_size)
+  input_ids = tf.keras.layers.Input(shape=(SEQUENCE_LENGTH,), name='input_ids', dtype=tf.int32)
+  input_mask = tf.keras.layers.Input(shape=(SEQUENCE_LENGTH,), name='input_mask', dtype=tf.int32)
+  label_ids = tf.keras.layers.Input(shape=(128,), name='input_type_ids', dtype=tf.int32)
```
Comment (saberkun) [inline rfcs/20200920-keras-nlp-bert.md]: If this is the mask token ids, it is not used inside this keras model?
How to use inside the loss?

File: rfcs/20200920-keras-nlp-bert.md [hunk 2]
```diff
@@ -0,0 +1,370 @@
...
+vocab_path = "gs://<bucket_name>/<file_path>/vocab.txt"
+tokenizer = tftext.BertTokenizer(vocab_path, token_out_type=tf.int64, lower_case=False)
+MAX_SEQ_LENGTH = 128
+def preprocess(input_text):
+  token_ids = tokenizer.tokenize(input_text)
+  cls = tf.expand_dims(tf.expand_dims(tf.cast(tf.tile([101], [token_ids.nrows()]), tf.int64), axis=1), axis=1)
+  sep = tf.expand_dims(tf.expand_dims(tf.cast(tf.tile([102], [token_ids.nrows()]), tf.int64), axis=1), axis=1)
+  token_ids = tf.concat([cls, token_ids, sep], axis=1)
+  segment_ids = tf.concat([tf.zeros_like(cls), tf.ones_like(token_ids), tf.ones_like(sep)], axis=1)
+  output_shape = [None, SEQUENCE_LENGTH]
```
Comment (nanhua2013) [inline rfcs/20200920-keras-nlp-bert.md:52]: Where is SEQUENCE_LENGTH defined? It's MAX_SEQ_LENGTH?

File: rfcs/20200920-keras-nlp-bert.md [hunk 3]
```diff
@@ -0,0 +1,370 @@
...
+with [ALBERT](https://arxiv.org/pdf/1909.11942.pdf), this can be achieved by:
+
+```python
+word_ids = tf.keras.layers.Input(shape=(None,), dtype=tf.int32, name='input_word_ids')
+mask = tf.keras.layers.Input(shape=(None,), dtype=tf.int32, name='input_mask')
+type_ids = tf.keras.layers.Input(shape=(None,), dtype=tf.int32, name='input_type_ids')
+embedding_layer = keras_nlp.layers.OnDeviceEmbedding(vocab_size, embedding_width)
+word_embeddings = embedding_layer(word_ids)
+position_embedding_layer = keras_nlp.layers.PositionEmbedding(max_sequence_length)
+position_embeddings = position_embedding_layer(word_embeddings)
```
Comment (nanhua2013) [inline rfcs/20200920-keras-nlp-bert.md]: nit: to be consistent with layers below, maybe we'd also anonymize embedding_layer and position_embedding_layer?
i.e. word_embeddings = keras_nlp.layers.OnDeviceEmbedding(vocab_size, embedding_width)(word_ids)
position_embeddings = keras_nlp.layers.PositionEmbedding(max_sequence_length)(word_embeddings)

File: rfcs/20200920-keras-nlp-bert.md [hunk 4]
```diff
@@ -0,0 +1,379 @@
...
+
+imdb_reviews = tfds.load('imdb_reviews')
+train_ds = imdb_reviews['train'].batch(32)
+test_ds = imdb_reviews['test'].batch(32)
+
+# Tokenization with BertTokenizer
+vocab_path = "gs://<bucket_name>/<file_path>/vocab.txt"
+tokenizer = tftext.BertTokenizer(vocab_path, token_out_type=tf.int64, lower_case=False)
+MAX_SEQ_LENGTH = 128
+def preprocess(input_text):
```
Comment (fchollet) [inline rfcs/20200920-keras-nlp-bert.md:49]: The preprocessing function here is quite difficult to parse. 101 and 102 are unexplained constants. There are many steps. We should simplify.

Question: why BertTokenizer instead of TextVectorization?

File: rfcs/20200920-keras-nlp-bert.md [hunk 5]
```diff
@@ -0,0 +1,379 @@
...
+      'input_mask': input_mask,
+      'input_type_ids': segment_ids
+  }
+
+strategy = tf.distribute.TPUStrategy(...)
+with strategy.scope():
+  encoder = keras_nlp.encoders.BertEncoder(vocab_size=30522, max_sequence_length=512, type_vocab_size=2)
+  checkpoint_path = "gs://<bucket_name>/<file_path>"
+  checkpoint = tf.train.Checkpoint(model=encoder)
+  checkpoint.restore(checkpoint_path).assert_consumed().run_restore_ops()
```
Comment (fchollet) [inline rfcs/20200920-keras-nlp-bert.md]: The 3 lines above (involving two classes and 3 methods) should just be `model.load_weights(checkpoint_path)`.

File: rfcs/20200920-keras-nlp-bert.md [hunk 6]
```diff
@@ -0,0 +1,379 @@
...
+    """
+```
+
+### Layers -- SelfAttentionMask
+
+```python
+class SelfAttentionMask(tf.keras.layers.Layer):
+  """Create 3D attention mask from a 2D tensor mask.
+
+    inputs[0]: from_tensor: 2D or 3D Tensor of shape
```
Comment (fchollet) [inline rfcs/20200920-keras-nlp-bert.md:218]: We should format this as a `Call arguments` section

File: rfcs/20200920-keras-nlp-bert.md [hunk 7]
```diff
@@ -0,0 +1,379 @@
...
+```
+
+## Detailed Design
+
+### Layers -- TransformerEncoderBlock
+
+This layer encapsulates a single layer of Transformer Encoder.
+
+```python
+class TransformerEncoderBlock(tf.keras.layers.Layer):
```
Comment (fchollet) [inline rfcs/20200920-keras-nlp-bert.md]: Why not `TransformerEncoder`?

File: rfcs/20200920-keras-nlp-bert.md [hunk 8]
```diff
@@ -0,0 +1,379 @@
...
+  def __init__(
+      self,
+      vocab_size,
+      hidden_size=768,
+      num_layers=12,
+      num_attention_heads=12,
+      max_sequence_length=512,
+      type_vocab_size=16,
+      inner_dim=3072,
+      inner_activation=lambda x: tf.keras.activations.gelu(x, approximate=True),
```
Comment (fchollet) [inline rfcs/20200920-keras-nlp-bert.md]: Default should be a string.

File: rfcs/20200920-keras-nlp-bert.md [hunk 9]
```diff
@@ -0,0 +1,379 @@
...
+      output_dropout=0.1,
+      attention_dropout=0.1,
+      initializer=tf.keras.initializers.TruncatedNormal(stddev=0.02),
```
Comment (fchollet) [inline rfcs/20200920-keras-nlp-bert.md]: Default should be a string. This also a surprising default value, why is this initializer not adaptable to the number of units?

File: rfcs/20200920-keras-nlp-bert.md (hunk 4 above)
Comment (alextp) [reply rfcs/20200920-keras-nlp-bert.md:49]: Or rather WordPieceVectorizer or some other name that describes in a high level what is actually happening here.

File: rfcs/20200920-keras-nlp-bert.md [hunk 10]
```diff
@@ -0,0 +1,379 @@
...
+  output = tf.keras.layers.Dense(1, activation='sigmoid')(x)
+  model = tf.keras.Model(inputs=[input_ids, input_mask, label_ids], outputs=output)
+
+model.compile('adam', 'binary_crossentropy', ['accuracy'])
+model.fit(train_ds, epochs=5, validation_data=test_ds)
+```
+
+### Pretraining task
+
+We aim to provide pretrained checkpoints for `BertEncoder`, however the user can choose to pretrain a new BertEncoder based on their
```
Comment (alextp) [inline rfcs/20200920-keras-nlp-bert.md]: Clarify what data these pre-trained checkpoints are trained on and their model sizes.

File: rfcs/20200920-keras-nlp-bert.md [hunk 11]
```diff
@@ -0,0 +1,379 @@
...
+
+  This layer uses either tf.gather or tf.one_hot to translate integer indices to
+  float embeddings.
+
+  Arguments:
+    vocab_size: Number of elements in the vocabulary.
+    embedding_width: Output size of the embedding layer.
+    initializer: The initializer to use for the embedding weights. Defaults to
+      "glorot_uniform".
+    use_one_hot: Whether to use tf.one_hot over tf.gather for the embedding
```
Comment (alextp) [inline rfcs/20200920-keras-nlp-bert.md:242]: This class feels like it's leaking implementation details very high up the user interface. I understand why it's necessary now, but:
 1. in the future we might be able to optimize matmul(x, one_hot(y)) to gather(x, y) for the cases where it makes sense
 2. in the future we might want to decide whether embeddings should live on the host or on device holistically based on things like the embedding size

Given these two things I think it makes sense to file bugs for the infra folks to investigate making these tradeoffs automatically and to call this class experimental to indicate that we hope to simplify it in the future.

File: rfcs/20200920-keras-nlp-bert.md [hunk 12]
```diff
@@ -0,0 +1,379 @@
...
+      lookup. Defaults to False (that is, using tf.gather). Setting this option
+      to True may improve performance, especially on small vocabulary sizes, but
+      will generally require more memory.
+    scale_factor: Whether to scale the output embeddings. Defaults to None (that
```
Comment (alextp) [inline rfcs/20200920-keras-nlp-bert.md]: Why do we need a scale factor instead of using another layer to scale?

File: rfcs/20200920-keras-nlp-bert.md [hunk 13]
```diff
@@ -0,0 +1,379 @@
...
+               initializer="glorot_uniform",
+               use_one_hot=False,
+               scale_factor=None,
+               **kwargs):
+```
+
+### Layers -- PositionEmbedding
+
+```python
+class PositionEmbedding(tf.keras.layers.Layer):
```
Comment (alextp) [inline rfcs/20200920-keras-nlp-bert.md:259]: There's more than one type of positional embedding in the literature (some people want them periodic, for example) so I think it's a good idea to call it something more specific.

Comment (saberkun) [reply rfcs/20200920-keras-nlp-bert.md:259]: @tanzhenyu The other common encoding is relative position encoding: https://github.com/tensorflow/models/blob/master/official/nlp/modeling/layers/position_embedding.py#L26
Note that, we need to consolidate the implementation related to: https://github.com/tensorflow/tensor2tensor/pull/177
TransformerXL implemented the paper version: https://github.com/tensorflow/models/blob/master/official/nlp/modeling/networks/xlnet_base.py#L341 

Thus, for the layer naming, maybe: PositionEmbedding->LearnedPositionEmbedding and the relative one is RelativePositionEmbedding/Encoding?

File: rfcs/20200920-keras-nlp-bert.md (hunk 12 above)
Comment (tanzhenyu) [reply rfcs/20200920-keras-nlp-bert.md]: I am leaning towards not having `scale_factor`, however it stems from a legacy implementation, which is used in the Transformer paper and maybe some T5 implementation. I think we should just let user do create a tf.keras.layers.Lambda to deal with it. 
@saberkun Agree?

File: rfcs/20200920-keras-nlp-bert.md (hunk 13 above)
Comment (tanzhenyu) [reply rfcs/20200920-keras-nlp-bert.md:259]: Hmmm...PositionEmbedding and PositionalEncoding are two different things. I believe when we say `Embedding`, it should always imply it's trainable, on the other hand `Encoding` implies it's not trainable. Thoughts?

File: rfcs/20200920-keras-nlp-bert.md (hunk 11 above)
Comment (tanzhenyu) [reply rfcs/20200920-keras-nlp-bert.md:242]: Makes sense. I filed a bug. 
On the other hand, do you think it's more appropriate if I claim the ideal solution is that this would be part of tf.keras.layers.Embedding, so no embedding layers are required here?

File: rfcs/20200920-keras-nlp-bert.md (hunk 10 above)
Comment (tanzhenyu) [reply rfcs/20200920-keras-nlp-bert.md]: Yes. For now I'm gonna direct this to TF Hub urls.

File: rfcs/20200920-keras-nlp-bert.md (hunk 9 above)
Comment (tanzhenyu) [reply rfcs/20200920-keras-nlp-bert.md]: Done.

File: rfcs/20200920-keras-nlp-bert.md (hunk 8 above)
Comment (tanzhenyu) [reply rfcs/20200920-keras-nlp-bert.md]: Done. However we should allow users to use the default activation, which is `approximate=True`, WDYT?

File: rfcs/20200920-keras-nlp-bert.md (hunk 7 above)
Comment (tanzhenyu) [reply rfcs/20200920-keras-nlp-bert.md]: I tried to rename it to be `TransformerEncoder`, but was pointed out that this is a single transformer layer and given Bert usually need to use 12 of them.
Second thought we should name this to be `TransformerEncoder` though. Updated.

Comment (tanzhenyu) [reply rfcs/20200920-keras-nlp-bert.md]: @saberkun FYI.

File: rfcs/20200920-keras-nlp-bert.md (hunk 5 above)
Comment (tanzhenyu) [reply rfcs/20200920-keras-nlp-bert.md]: Done

File: rfcs/20200920-keras-nlp-bert.md (hunk 6 above)
Comment (tanzhenyu) [reply rfcs/20200920-keras-nlp-bert.md:218]: Done

File: rfcs/20200920-keras-nlp-bert.md (hunk 4 above)
Comment (tanzhenyu) [reply rfcs/20200920-keras-nlp-bert.md:49]: Re Francois: yes 101 and 102, I updated to use `tokenize_with_offsets` instead, which will do the prepend and append logic.
BertTokenizer is used to tokenize words into sub-words, which is the word piece algorithm.

Re Alex: BertTokenizer = BasicTokenizer (uncasing, split, standardization) + WordPieceTokenizer. This is a mirroring of the current status of tf.text ops. I do agree once we have Tokenization RFC, we should name this to be keras_nlp.ops.WordPieceTokenizer that performs both. For now I'd like this example can actually be run successfully in colab.

File: rfcs/20200920-keras-nlp-bert.md (hunk 3 above)
Comment (tanzhenyu) [reply rfcs/20200920-keras-nlp-bert.md]: Done.

File: rfcs/20200920-keras-nlp-bert.md (hunk 2 above)
Comment (tanzhenyu) [reply rfcs/20200920-keras-nlp-bert.md:52]: Fixed.

File: rfcs/20200920-keras-nlp-bert.md (hunk 1 above)
Comment (tanzhenyu) [reply rfcs/20200920-keras-nlp-bert.md]: Yeah renamed this to type_id instead.

File: rfcs/20200920-keras-nlp-bert.md (hunk 4 above)
Comment (YunusBulut) [reply rfcs/20200920-keras-nlp-bert.md:49]: Not very crucial but there's typo in the comment: "Considering a imbd review dataset". It should be "Considering an imdb review dataset".

File: rfcs/20200920-keras-nlp-bert.md [hunk 14]
```diff
@@ -0,0 +1,370 @@
...
+
+This proposal builds on the assumption that inputs are lookup indices, i.e., `tf.int64` sequences.
+Tokenization is not part of this proposal but will be our immediate next step.
+
+### Classification task
+
+Case where a user want to use a pretrained BERT encoder for sentiment analysis:
+
+```python
+# Considering a imbd review dataset
```
Comment (YunusBulut) [inline rfcs/20200920-keras-nlp-bert.md:35]: Not very crucial but there's typo in the comment: "Considering a imbd review dataset". It should be "Considering an imdb review dataset".
